src/cogent3/maths/distance_transform.py

```diff
@@ -1253,7 +1267,41 @@ def binary_dist_hamming(datamtx, strict=True):
     return dists
 
 
-def binary_dist_jaccard(datamtx, strict=True):
+def fast_jaccard(x: iter, y: iter):
+    xset = set(x)
+    yset = set(y)
+    return 1 - len(xset & yset) / len(xset | yset)
+
+
+@singledispatch
+def jaccard(x, y):
```

Comment:
fully type hint signature
```python
def jaccard(x: T, y: T) -> float:
```

Reply:
I think the only problem with this will be the implication that the function allows mixed types, eg: calculating the difference between a set and a NDArray, or vice versa.  And now I think of it, the current implementation won't handle calculating the difference between a set and a frozenset.  I suspect your deceptively not-recursive suggestion of casting to sets then calling the set,set function might handle this.  I'll have a play with it and see if I can't make something elegant do the job.

Reply:
`set()` and `frozenset()` instances can operate together. So no need to worry about that mixture.

I get what you mean about the type hint. I wonder if code editors like VS Code, PyCharm, will infer that `x` and `y` should have the same type and provide feedback to user when their usage doesn't match.

More importantly, we need to set limits to what we should try and handle.

For instance, there are limits to the `singledispatch` mechanism, mainly that it delegates based solely on the type of the first argument. For example, if a user invokes jaccard with `type(x)==set` and `type(y)==ndarray`, the set variant of `jaccard` will be invoked, and fail.

My view on cases such is that this reflects a fundamental error in the users code. Better to allow an exception happen and let them identify that error.

Reply:
I tried probing singledispatch to verify that, and it's behaviour is perhaps even more disturbing.  If you have multiple prototypes with the same type for the first parameter, the last one is chosen.  I assume the internal data structure is a dict with the type of the first parameter as the key and a pointer to the function as the value, and subsequent implementations collide and overwrite the entry for the first parameter type.  That design is likely to cause developers relying upon duck typing magic to get into trouble.

I addressed the mixed type issue that can't be resolved programmatically using singledispatch by altering the doc comment

Reply:
Document is the right way to go!